Fix constraints and constraint test files per issue #773

Rene2mt opened the pull request:
# Committer Notes

This PR makes constraint, baseline, and example content (a.k.a. "template") updates to resolve issue https://github.com/GSA/fedramp-automation/issues/773

Specifically, it makes the following changes:

- For FedRAMP extension properties, changes namespace from "**https**://fedramp.gov/ns/oscal" to "**http**://fedramp.gov/ns/oscal"
- For the SSP `identifier-type`, changes the value from "**https**://fedramp.gov/ns/oscal" to "**http**://fedramp.gov/ns/oscal"
- For POA&M `facets`, changes the `system` flag from "**https**://fedramp.gov" to "**http**://fedramp.gov/ns/oscal"

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [x] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

DimitriZhurkin commented:
Sorry to chime in quite late. I think all files in the latest branch already have instances of `http:/` replaced with `https:/`.

Apologies if I'm off the track.

Gabeblis commented:
Is the goal to update every occurrence of `ns="https..."` in the repo? If so, this seems a little light

aj-stein-gsa commented:
> Any chance we can just use **gov.fedramp.oscal** namespace? Makes more sense to me

I mean could we? Sure. Is it valid within the spec of URIs? Not quite, especially when you consider we have a full URI with a subpath, so you are proposing `gov.fedramp.oscal/ns/oscal` in full?

Technically speaking (because we are that kind of group), reverse DNS notation is not how an actual domain works in a URI, more details [in the official RFC about the host requirement](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2).

wandmagic commented:
> > Any chance we can just use **gov.fedramp.oscal** namespace? Makes more sense to me
> 
> I mean could we? Sure. Is it valid within the spec of URIs? Not quite, especially when you consider we have a full URI with a subpath, so you are proposing `gov.fedramp.oscal/ns/oscal` in full?
> 
> Technically speaking (because we are that kind of group), reverse DNS notation is not how an actual domain works in a URI, more details [in the official RFC about the host requirement](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2).

then could we just have `fedramp.gov/ns/oscal`? the confusing part for me is when i see http i expect to be able to send traffic to that domain over http protocol

aj-stein-gsa commented:
I would check your environment and/or use the container we provide, I am unable to get the extraneous error anymore. Please cross-check not just the branches, but also the current commit (`git rev-parse HEAD`) for our repos.

```sh
me@computer telos-fedramp-pilot % oscal-cli --version
oscal-cli 2.4.0 built at 2024-11-26 17:07 from branch dba6d9c570f0aa42022d9754df42d1dc5fc295d4 (dba6d9c) at https://github.com/metaschema-framework/oscal-cli
liboscal-java  built at 2024-11-26 16:40 from branch 2f3a394fa856e2bc90b74425c639c9bc107ea4e6 (2f3a394) at https://github.com/metaschema-framework/liboscal-java
oscal v1.1.3 built at 2024-11-26 16:40 from branch b123c11bd12c8b8f1bcc8bf85763e5775c0423e9 (b123c11) at https://github.com/usnistgov/OSCAL.git
metaschema-java 2.1.0 built at 2024-11-26T16:21:47+0000 from branch 462da0c64add5b369af740f4d2057643ac72b056 (462da0c) at https://github.com/metaschema-framework/metaschema-java
metaschema 2.1.0 built at 2024-11-26T16:21:47+0000 from branch b6601f7430f83f1a53a11bf32575b69e131bc912 (b6601f7) at https://github.com/metaschema-framework/metaschema.git
me@computer telos-fedramp-pilot % git branch --show-current
main
me@computer telos-fedramp-pilot % git rev-parse HEAD
82e80b82d9908857f44cb95949859262c4e5cc82
me@computer telos-fedramp-pilot % pushd ~/Code/fedramp-automation
~/Code/fedramp-automation ~/Code/telos-fedramp-pilot
me@computer fedramp-automation % git branch --show-current      
develop
me@computer fedramp-automation % git rev-parse HEAD             
48dd7f54ced994a03e5e455e545d686dc814c5c5
me@computer fedramp-automation % oscal-cli validate '/home/me/telos-fedramp-pilot/OSCAL SSP/Hogwarts SSP - 1.0 (2024-11-19T093555Z).json' -c '/home/me/fedramp-automation/src/validations/constraints/fedramp-external-allowed-values.xml' -c '/home/me/fedramp-automation/src/validations/constraints/fedramp-external-constraints.xml' -c '/home/me/fedramp-automation/src/validations/constraints/oscal-external-constraints.xml'
Loading 'file:/home/me/fedramp-automation/src/validations/constraints/fedramp-external-allowed-values.xml'
Loading 'file:/home/me/fedramp-automation/src/validations/constraints/fedramp-external-constraints.xml'
Loading 'file:/home/me/fedramp-automation/src/validations/constraints/oscal-external-constraints.xml'
Validating 'file:///home/me/telos-fedramp-pilot/OSCAL%20SSP/Hogwarts%20SSP%20-%201.0%20(2024-11-19T093555Z).json' as JSON.
me@computer fedramp-automation % oscal-cli validate '//home/me/telos-fedramp-pilot/OSCAL SSP/Hogwarts SSP - 1.0 (2024-11-19T093555Z).json' -c '//home/me/fedramp-automation/src/validations/constraints/fedramp-external-allowed-values.xml' -c '//home/me/fedramp-automation/src/validations/constraints/fedramp-external-constraints.xml' -c '//home/me/fedramp-automation/src/validations/constraints/oscal-external-constraints.xml'
Loading 'file://home/me/fedramp-automation/src/validations/constraints/fedramp-external-allowed-values.xml'
Loading 'file://home/me/fedramp-automation/src/validations/constraints/fedramp-external-constraints.xml'
Loading 'file://home/me/fedramp-automation/src/validations/constraints/oscal-external-constraints.xml'
Validating 'file:////home/me/telos-fedramp-pilot/OSCAL%20SSP/Hogwarts%20SSP%20-%201.0%20(2024-11-19T093555Z).json' as JSON.
Validation identified the following issues:
[ERROR] [/system-security-plan/system-implementation[1]/component[7]] component-has-authentication-method: A FedRAMP SSP MUST include at least one authentication method for each leveraged system.
[ERROR] [/system-security-plan/system-implementation[1]/component[8]] component-has-authentication-method: A FedRAMP SSP MUST include at least one authentication method for each leveraged system.
[ERROR] [/system-security-plan/system-implementation[1]/component[7]] component-has-non-provider-responsible-role: A FedRAMP SSP MUST have each component describing leveraged systems, interconnections, or authorized services identify at least one responsible role other than "provider".
[ERROR] [/system-security-plan/system-implementation[1]/component[8]] component-has-non-provider-responsible-role: A FedRAMP SSP MUST have each component describing leveraged systems, interconnections, or authorized services identify at least one responsible role other than "provider".
[ERROR] [/system-security-plan/system-implementation[1]/component[7]] component-has-provider-responsible-role: A FedRAMP SSP MUST have each component describing leveraged systems, interconnections, or authorized services identify a "provider" role that references one responsible party.
[ERROR] [/system-security-plan/system-implementation[1]/component[8]] component-has-provider-responsible-role: A FedRAMP SSP MUST have each component describing leveraged systems, interconnections, or authorized services identify a "provider" role that references one responsible party.
[ERROR] [/system-security-plan/system-implementation[1]/leveraged-authorization[1]/prop[3]/@value] leveraged-authorization-has-valid-impact-level: A FedRAMP SSP MUST define the appropriate FIPS-199 impact level (low, moderate, or high) for each leveraged authorization.
[ERROR] [/system-security-plan] non-provider-responsible-role-references-user: A FedRAMP SSP MUST have each component describing leveraged systems, interconnections, or authorized services reference at least one user with an authorized privilege and function performed via the "privilege-uuid" property.
[ERROR] [/system-security-plan/metadata[1]] data-center-alternate: There MUST be one or more alternate data center(s).
[ERROR] [/system-security-plan/metadata[1]] data-center-count: There MUST be at least two (2) data centers listed.
[ERROR] [/system-security-plan/metadata[1]] data-center-primary: There MUST be a single primary data center.
[ERROR] [/system-security-plan/metadata[1]] responsible-party-prepared-by-location-valid: A FedRAMP SSP MUST have a responsible party for preparing the document, and that party MUST define an address.
[ERROR] [/system-security-plan/metadata[1]] fedramp-version: A FedRAMP document's metadata MUST define a valid FedRAMP version.
[ERROR] [/system-security-plan/metadata[1]] marking: A FedRAMP document MUST have a marking that defines its data classification.
[WARNING] [/system-security-plan/metadata[1]/party[3]/telephone-number[1]] Value '1800-123-4567' did not match the pattern '^[0-9]{3}[0-9]{1,12}$' at path '/system-security-plan/metadata[1]/party[3]/telephone-number[1]'
[WARNING] [/system-security-plan/metadata[1]/party[5]/telephone-number[1]] Value '001-46785-115547' did not match the pattern '^[0-9]{3}[0-9]{1,12}$' at path '/system-security-plan/metadata[1]/party[5]/telephone-number[1]'
[WARNING] [/system-security-plan/metadata[1]/party[8]/telephone-number[1]] Value '12345678522-2' did not match the pattern '^[0-9]{3}[0-9]{1,12}$' at path '/system-security-plan/metadata[1]/party[8]/telephone-number[1]'
[WARNING] [/system-security-plan/metadata[1]/party[10]/telephone-number[1]] Value '123456789-4452' did not match the pattern '^[0-9]{3}[0-9]{1,12}$' at path '/system-security-plan/metadata[1]/party[10]/telephone-number[1]'
[ERROR] [/system-security-plan/system-characteristics[1]] has-cloud-service-model-remarks: A FedRAMP SSP with a cloud service model of "other" MUST supply remarks to explain this choice.
[ERROR] [/system-security-plan/system-characteristics[1]] has-fully-operational-date: A FedRAMP SSP MUST define the system's fully operational date.
[ERROR] [/system-security-plan/system-characteristics[1]] has-system-id: A FedRAMP SSP MUST have a FedRAMP system identifier.
[ERROR] [/system-security-plan/system-characteristics[1]/authorization-boundary[1]/diagram[1]/link[1]] has-authorization-boundary-diagram-link-href-target: A FedRAMP SSP MUST include an authorization boundary diagram.
[ERROR] [/system-security-plan/system-characteristics[1]/network-architecture[1]/diagram[1]/link[1]] has-network-architecture-diagram-link-href-target: A FedRAMP SSP MUST include a network architecture diagram.
[ERROR] [/system-security-plan/system-characteristics[1]/data-flow[1]/diagram[1]/link[1]] has-data-flow-diagram-link-href-target: A FedRAMP SSP MUST include a data flow diagram.
[ERROR] [/system-security-plan/system-characteristics[1]/data-flow[1]/diagram[2]/link[1]] has-data-flow-diagram-link-href-target: A FedRAMP SSP MUST include a data flow diagram.
[ERROR] [/system-security-plan/system-implementation[1]] has-inventory-items: A FedRAMP SSP system implementation section MUST have at least two inventory items.
[ERROR] [/system-security-plan/system-implementation[1]/component[8]] inter-boundary-component-direction-incoming-has-ipv-uri: Component a0d10f42-41eb-5407-97ae-bd35b3d2e455 (/system-security-plan/system-implementation[1]/component[8]) MUST have at least one local ipv4 address, ipv6 address, or a URI to an API.
[ERROR] [/system-security-plan/system-implementation[1]/component[8]] network-component-has-connection-security-prop: All network components in a FedRAMP SSP system implementation MUST define at least one interconnection security property.
[ERROR] [/system-security-plan/system-implementation[1]/component[9]] network-component-has-implementation-point: A FedRAMP SSP with service components and CLI software components performing cross-boundary network communication MUST indicate exactly one time if the point of implementation is internal or external to the system.
[ERROR] [/system-security-plan/system-implementation[1]/component[10]] network-component-has-implementation-point: A FedRAMP SSP with service components and CLI software components performing cross-boundary network communication MUST indicate exactly one time if the point of implementation is internal or external to the system.
[ERROR] [/system-security-plan/system-implementation[1]/component[11]] network-component-has-implementation-point: A FedRAMP SSP with service components and CLI software components performing cross-boundary network communication MUST indicate exactly one time if the point of implementation is internal or external to the system.
[ERROR] [/system-security-plan/system-implementation[1]/component[12]] network-component-has-implementation-point: A FedRAMP SSP with service components and CLI software components performing cross-boundary network communication MUST indicate exactly one time if the point of implementation is internal or external to the system.
[ERROR] [/system-security-plan/system-implementation[1]/component[8]/prop[5]/@value] external-system-nature-of-agreement: Value 'Contract' doesn't match one of 'contract, eula, isa, license, mou, other, or sla' at path '/system-security-plan/system-implementation[1]/component[8]/prop[5]/@value'
[WARNING] [/system-security-plan/system-implementation[1]/component[10]/protocol[1]] It is a best practice to provide a UUID.
[WARNING] [/system-security-plan/system-implementation[1]/component[11]/protocol[1]] It is a best practice to provide a UUID.
[WARNING] [/system-security-plan/system-implementation[1]/component[12]/protocol[1]] It is a best practice to provide a UUID.
[ERROR] [/system-security-plan/control-implementation[1]/implemented-requirement[82]/statement[1]] missing-response-components: A FedRAMP SSP MUST identify how the system implements each control requirement implemented at the per-statement level and reference any component used to implement it.
[ERROR] [/system-security-plan/control-implementation[1]/implemented-requirement[82]/statement[2]] missing-response-components: A FedRAMP SSP MUST identify how the system implements each control requirement implemented at the per-statement level and reference any component used to implement it.
[ERROR] [/system-security-plan/control-implementation[1]/implemented-requirement[82]] The cardinality '0' is below the required minimum '1' for items matching './/by-component'.
[ERROR] [/system-security-plan/back-matter[1]] has-fedramp-citations: A FedRAMP MUST be have exactly one resource with a link to the FedRAMP Laws, Regulations, Standards and Guidance, but 0 found.
[ERROR] [/system-security-plan/back-matter[1]/resource[6]/prop[1]/@value] Value 'separation-of-duties-matrix' doesn't match one of 'acronyms, administrators-guide, agreement, artifact, citation, evidence, external-guidance, image, interview-notes, law, logo, plan, policy, procedure, questionnaire, raw-data, regulation, report, rules-of-behavior, screen-shot, standard, system-guide, tool-output, or users-guide' at path '/system-security-plan/back-matter[1]/resource[6]/prop[1]/@value'
The file 'file:////home/me/telos-fedramp-pilot/OSCAL%20SSP/Hogwarts%20SSP%20-%201.0%20(2024-11-19T093555Z).json' is invalid.
me@computer fedramp-automation % oscal-cli validate '//home/me/telos-fedramp-pilot/OSCAL SSP/Hogwarts SSP - 1.0 (2024-11-19T093555Z).json' -c '//home/me/fedramp-automation/src/validations/constraints/fedramp-external-allowed-values.xml' -c '//home/me/fedramp-automation/src/validations/constraints/fedramp-external-constraints.xml' -c '//home/me/fedramp-automation/src/validations/constraints/oscal-external-constraints.xml' --no-color 2>&1| grep --only-matching --color -i extraneous
# No output returned
```

aj-stein-gsa commented on src/content/rev5/examples/ssp/xml/fedramp-ssp-example.oscal.xml:
Blocking: these changes are outside the range of the diff but can you please correc them?

https://github.com/Rene2mt/fedramp-automation/blob/b244ff0e0b3ba0ae31bc21d175c8052461823bd4/src/content/rev5/examples/ssp/xml/fedramp-ssp-example.oscal.xml#L1201-L1204

https://github.com/Rene2mt/fedramp-automation/blob/b244ff0e0b3ba0ae31bc21d175c8052461823bd4/src/content/rev5/examples/ssp/xml/fedramp-ssp-example.oscal.xml#L1793-L1796

https://github.com/Rene2mt/fedramp-automation/blob/b244ff0e0b3ba0ae31bc21d175c8052461823bd4/src/content/rev5/examples/ssp/xml/fedramp-ssp-example.oscal.xml#L2211-L2214

Rene2mt replied:
fixed in commit 8ca713b5586832e77f6106b976956b132d40e348

aj-stein-gsa commented on src/content/rev5/templates/sap/xml/FedRAMP-SAP-OSCAL-Template.xml:
```diff
@@ -814,7 +814,7 @@
          <!-- Section 6.3 Testing performed using automated tools -->
          <part ns="https://fedramp.gov/ns/osccal" name="automated-tools-testing">
```
Blocking: this still needs to be changed.

```suggestion
         <part ns="http://fedramp.gov/ns/osccal" name="automated-tools-testing">
```

aj-stein-gsa commented on src/content/rev5/templates/sar/xml/FedRAMP-SAR-OSCAL-Template.xml:
Blocking: I thought we talked about clean up of the `facet/@system`s as well?

https://github.com/Rene2mt/fedramp-automation/blob/b244ff0e0b3ba0ae31bc21d175c8052461823bd4/src/content/rev5/templates/sar/xml/FedRAMP-SAR-OSCAL-Template.xml#L942-L947

Note, I found a few of those, this is just the first sequence.

Rene2mt replied:
commit 8ca713b5586832e77f6106b976956b132d40e348

aj-stein-gsa commented on src/validations/constraints/content/ssp-interconnection-security-VALID.xml:
Blocking: are we going to correce the `system-id/@identifier-type` or is that out of scope for now? I can't tell 100% from the ADR and previous discussion.

https://github.com/Rene2mt/fedramp-automation/blob/b244ff0e0b3ba0ae31bc21d175c8052461823bd4/src/validations/constraints/content/ssp-interconnection-security-VALID.xml#L165

Rene2mt replied:
fixed in commit 22365169f44d6e1e25664de0ef423ee2cc7e8b78

aj-stein-gsa commented on src/validations/constraints/content/ssp-interconnection-security-INVALID.xml:
Blocking: ditto here on `system-id/@identifier-type`, did we plan on correcting this one?

aj-stein-gsa commented on src/validations/constraints/content/ssp-has-network-architecture-diagram-link-href-target-VALID-1.xml:
There are some `prop`s that were missed.

https://github.com/Rene2mt/fedramp-automation/blob/b244ff0e0b3ba0ae31bc21d175c8052461823bd4/src/validations/constraints/content/ssp-has-network-architecture-diagram-link-href-target-VALID-1.xml#L234-L235

Rene2mt replied:
fixed in commit 22365169f44d6e1e25664de0ef423ee2cc7e8b78

aj-stein-gsa reviewed (changes requested):
Nice work, but I found a few discrepancies we need to clean up before it is ready for merge and we send a notice to the community. I really appreciate this kind of tedious yeoman's task, so thanks for putting in the effort, sir.

Rene2mt commented on src/content/rev5/examples/ssp/xml/fedramp-ssp-example.oscal.xml:
```diff
@@ -561,7 +561,8 @@
 
    </metadata>
-   <import-profile href="https://raw.githubusercontent.com/GSA/fedramp-automation/refs/heads/master/dist/content/rev5/baselines/xml/FedRAMP_rev5_MODERATE-baseline_profile.xml">
+   <!-- <import-profile href="https://raw.githubusercontent.com/GSA/fedramp-automation/refs/heads/master/dist/content/rev5/baselines/xml/FedRAMP_rev5_MODERATE-baseline_profile.xml"> -->
```
fixed in commit 8ca713b5586832e77f6106b976956b132d40e348

Rene2mt commented:
@aj-stein-gsa good catches on the review.  I must have missed some of the more recent constraints when rebasing and also missed a couple if `identifier-type` flags.  I fixed the items you found and also updated the "AwesomeCloud" SSPs for good measure (although I wasn't sure if those are going away now that we have a more robust example SSP).  

So now, global search for "https://fedramp.gov" in vscode (excluding vendor, src\content\rev5\resources\xml folders) is only turning up ADR11 and cases where there are links to the FedRAMP website.

aj-stein-gsa commented:
If you are a community member subscribed to notifications outside of the developer team, be advised we are readying this for merge into `develop`. An email will be sent to pilot participants as well when this is complete, but this is a heads up that process is starting.
